### PR TITLE
add bitops module for optimized bit manipulation.

### PR DESCRIPTION
**Pull request by Parashurama**
This module implements reasonably optimized proc for bit manipulation/operation and optionally make uses of compiler intrinsics for improved performance.
It also support non-compiler backend with default implementation

### CODE REVIEW

File: lib/pure/bitopts.nim [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+      ## Counts the set bits in integer. (also called Hamming weight.)
+      var v = uint64(n)
+      v = v - ((v shr 1'u64) and 0x5555555555555555'u64)
+      v = (v and 0x3333333333333333'u64) + ((v shr 2'u64) and 0x3333333333333333'u64)
+      v = (v + (v shr 4'u64) and 0x0F0F0F0F0F0F0F0F'u64)
+      result = ((v * 0x0101010101010101'u64) shr 56'u64).cint
+
+
+proc rotateLeftBits*(value: uint8;
+                     amount: static[range[1..7]]): uint8 {.inline, noSideEffect.} =
```
Comment (Araq) [inline lib/pure/bitopts.nim]: Use 'range[1..7]' without the 'static' please, likewise for the following usages.

File: lib/pure/bitopts.nim [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+#        (c) Copyright 2017 Nim Authors
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+#
+
+## This module implements a series of low level methods for bit manipulation.
+## By default, this module use fairly optimized procs written in pure nim.
+##
+## To improve performance, you can use the `builtinBitOpts` flag. This unlock
```
Comment (Araq) [inline lib/pure/bitopts.nim]: I think this should be the default. And I'm afraid the implementations should use internally the ``when nimvm`` feature to provide the pure Nim implementation for compiletime evaluation purposes.

**Comment by Araq** [comment]
A great job and worth adding to the stdlib. @dom96 what do you think?

File: lib/pure/bitopts.nim (hunk 1 above)
Comment (Parashurama) [inline lib/pure/bitopts.nim]: Done

File: lib/pure/bitopts.nim (hunk 2 above)
Comment (Parashurama) [inline lib/pure/bitopts.nim]: Will switch to compiler intrinsics by default.
Actually, I think it already works for compile time.
see: static block at the end of  tests/stdlib/tbitopts.nim
but I might be wrong

**Comment by Araq** [comment]
Test "tests/stdlib/tbitopts.nim" in category "stdlib"
Failure: reNimcCrash
Expected:
Gotten:
execution of an external compiler program 'clang -c  -w  -Ilib -o tests/stdlib/nimcache/compiler_tbitopts.o tests/stdlib/nimcache/compiler_tbitopts.c' failed with exit code: 1

**Comment by Parashurama** [comment]
Yeah, apparently the reference page I used to generate importc proc had a few errors, so it crashes on clang generated c code.
I also cannot check at the moment whether VCC or ICC code path compiles correctly, having no windows installed.
I can try to install Intel compiler on Debian.

**Comment by mfxmfx** [comment]
I agree that this would be a useful module, but I think the name should be "bitops" and not "bitopts".

Furthermore it should be "countLeadingZeroBits" instead of "leadingZeroBits" - maybe you could do some quick research how other tools and libraries (e.g. gcc, llvm) and languages (C++, Go, D, Rust) do name those operations so that we better follow existing practice.

**Comment by Parashurama** [comment]
I wouldn't mind renaming the module but let's see what people think.
I would also welcome people testing the code especially on windows platform, with mingw, visual c++ and Intel compiler, since I don't have access to those.

**Comment by andreaferretti** [comment]
By the way, I think it would be useful to add  `popcount` as an alias for `countSetBits`, since the same is pretty standard in literature.

In any case, this PR is a very welcome addition!

**Comment by dom96** [comment]
Yes, this is a great PR! I agree with @mfxmfx and I also think that you need to modify your code to follow our [style guide](https://github.com/nim-lang/Nim/wiki/Style-Guide-for-Nim-Code), i.e. use 2 spaces for indentation instead of 4. Should be an easy fix though.

**Comment by cheatfate** [comment]
@Parashurama it would be nice to have `bswap` intrinsics too.
I have made and tested some code with them, so you can more easily adopt it to your module
https://github.com/cheatfate/asyncpg/blob/master/asyncpg/byteswap.nim.

**Comment by Parashurama** [comment]
Actually I recently added bswap support  in lib/pure/endians.nim but it uses the previous interface with
```swapEndiansXX(dst: pointer, src: pointer)```
which is a bit cumbersome when swapping integers.
It may be a good idea to add bswap to bitops too.

**Comment by Araq** [comment]
> I wouldn't mind renaming the module or the procs, but let's see what people think.

The module name should be ``bitops`` indeed.

**Comment by Parashurama** [comment]
I made some modifications as requested. I think @cheatfate request is a good idea but it probably needs its own pull request in order to integrate ``bitopts.nim`` to ``endians.nim``.

**Comment by jangko** [comment]
I have tested with mingw-gcc 5.3.0-64bit and msvc-2015-64bit
it pass both tbitops and tbitops2.
failed to compile using msvc-2010 because standard msvc-2010 only support 32bit mode and need additional sdk to compile 64bit code.
seems the 32bit and 64bit bitops compilation should be separated with compile-time check.

**Comment by Parashurama** [comment]
Thanks, I figured there was something like that on windows 32 bits.
In the docs for MSVC and ICC, some intrinsics only exists on 64bits (ie popcnt64 and friends).

To make things easier, I will fall-back to pure nim implementation for 64bits integers on windows 32bits.

**Comment by jangko** [comment]
mingw-gcc-5.3.0-64bit: ok
mingw-gcc-4.8.1-32bit: ok
mingw-gcc-6.3.0-32bit: ok
mingw-gcc-6.3.0-64bit: ok

msvc-2015-64bit: failed
msvc-2015-32bit: failed
msvc-2010-32bit: failed

```text
reason:
  line 130: _bitScanForward    should be  _BitScanForward
  line 131: _bitScanForward64  should be  _BitScanForward64

failed test for msvc:
  doAssert( U64A.fastLog2 == 62) # ok in 32-bit mode
  doAssert( I64A.fastLog2 == 62) # ok in 32-bit mode
  doAssert( U32.fastLog2 == 31)
  doAssert( I32.fastLog2 == 31)
  doAssert( U16.fastLog2 == 13)
  doAssert( I16.fastLog2 == 13)
  doAssert( U8.fastLog2 == 5)
  doAssert( I8.fastLog2 == 5)
  
possible solution:
line 238 in fastLog2:
  result = 31 - vcc_scan_impl(bitScanReverse, x.culong)
ok when I change to(not sure):
  result = vcc_scan_impl(bitScanReverse, x.culong)
          
line 240 in fastLog2:
  result = 63 - vcc_scan_impl(bitScanReverse64, x.uint64)
ok when I change to(not sure):
  result = vcc_scan_impl(bitScanReverse64, x.uint64)
```

**Comment by Parashurama** [comment]
This should be fixed now.
The algorithm for fast_log2  is different in gcc/clang from the others.
for gcc/clang i use log2 = sizeof(integer) - 1 - numberleadingZero
for VCC/ICC  log2 = indexOfLastSetbit.

**Comment by markus-oberhumer** [comment]
I can help with adding support for more exotic compilers like icc on Linux and suncc, but we need to get the basics right first:

- countLeadingZeros & friends must return a sensible value for 0

- rotateLeft & friends should allow `SomeInteger` as second parameter and add an extra "and" to
mask the relevant bits (see `nimNewShiftOps` in system.nim)

- make sure everything works in Javascript (missing 64-bit integers)

FYI, here is a link to a [related discussion for Go](https://github.com/golang/go/issues/18616).

**Comment by Parashurama** [comment]
That's debatable if a sensible value for 0 is required. Most implementation are undefined for performance reasons.
I don't know anything but the basics about javascript. Apparently numbers are always double precision?

**Comment by jangko** [comment]
TDM-gcc-5.3.0-64bit: ok
mingw32-gcc-4.8.1-32bit: ok
mingw-w64-gcc-6.3.0-32bit: ok
mingw-w64-gcc-6.3.0-64bit: ok

msvc-2015-64bit: ok
msvc-2015-32bit: ok
msvc-2010-32bit: ok

```text
but i still found issues with icc:
  the first letter should be lower case or failed to compile:
    line 146: proc BitScanReverse -> proc bitScanReverse
    line 147: proc BitScanReverse64 -> proc bitScanReverse64

  builtin_popcnt32 expected a cint not uint32:
    line 171: builtin_popcnt32(x.uint32).int -> builtin_popcnt32(x.cint).int
```

**Comment by Parashurama** [comment]
Ok, thanks for your help. This should work now, hopefully :)
I now use this file https://gist.github.com/Parashurama/5146660e4915c48f491774a83d1967d2#file-exec_testbitops-sh for testing
I don't have either MSVC or Intel compiler, but at least the C code is generated and should avoid nim source errors.

**Comment by andreaferretti** [comment]
Any news on when this will be merged?

**Comment by markus-oberhumer** [comment]
IMO this currently does not match the quality required for a "standard library"

- undefined behaviour for zero values (even Rust and Go do get this right)

- missing Javascript support

- dubious rotateLeft & friends params (see nimNewShiftOps)

**Comment by andreaferretti** [comment]
I think that people that need to access the C intrinsics don't want to pay the additional cost of branching to check for zero values.

About Javascript: I see that there are pure nim implementations as fallback - I have not tried them, but I do not see why they should not work

**Comment by Parashurama** [comment]
That was my initial thought. If you are using this module, you likely looking for performance don't want to pay the additional cost of branching to check for zero values.
I could add an assert in the relevant procs and a compile flag to optionally support zero values, and avoid undefined behavior.

Javascript & nimvm should use the pure nim versions of procs. Although not actually tested with Javascript.

for rotateLeft&Right: shift values bigger than integer size are undefined. They checked in debug mode.

**Comment by Parashurama** [comment]
Okay, I added compile a time flag  'noUndefinedBitOpts' to guarantee no that no procs will trigger undefined behavior. It does incur a small performance penalty from zero branching.

I also found a rotateBits implementation that can handle 0 and integer size bitcount rotation.
see: https://en.wikipedia.org/wiki/Circular_shift#Implementing_circular_shifts
I checked x64 assembly output  for clang/gcc and it appear to compile to a single ROR/ROL instruction (in release mode, of course)

**Comment by Parashurama** [comment]
Okay, I added compile a time flag  'noUndefinedBitOpts' to guarantee no that no procs will trigger undefined behavior. It does incur a small performance penalty from zero branching.

I also found a rotateBits implementation that can handle 0 and integer size bitcount rotation.
see: https://en.wikipedia.org/wiki/Circular_shift#Implementing_circular_shifts
I checked x64 assembly output  for clang/gcc and it appear to compile to a single ROR/ROL instruction (in release mode, of course)

**Comment by andreaferretti** [comment]
Is this landing soon? Right now I am using it just by copying the definitions in my code, but it would be very nice to have this in the stdlib

**Comment by Parashurama** [comment]
I think it's ready to be merged now. If there is any problem, it can always be corrected with a PR.

File: lib/pure/bitops.nim [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+  when nimvm:
+    result = firstSetBit(x) - 1
+  else:
+    when useGCC_builtins:
+      when sizeof(x) <= 4: result = builtin_ctz(x.uint32).int
+      else:                result = builtin_ctzll(x.uint64).int
+    else:
+      result = firstSetBit(x) - 1
+
+{.push checks: off}
```
Comment (Araq) [inline lib/pure/bitops.nim]: wait, what? shouldn't the code work with Nim's runtime checks?

File: lib/pure/bitops.nim [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+  # Returns the number of leading 0-bits in x, starting at the most significant bit position. If x is 0, the result is undefined.
+  proc bitScanReverse(p: ptr uint32, b: uint32): cuchar {.importc: "_BitScanReverse", header: "<immintrin.h>", nosideeffect.}
+  proc bitScanReverse64(p: ptr uint32, b: uint64): cuchar {.importc: "_BitScanReverse64", header: "<immintrin.h>", nosideeffect.}
+
+  template icc_scan_impl(fnc: untyped; v: untyped): int =
+    var index: uint32
+    discard fnc(index.addr, v)
+    index.int
+
+{.push rangeChecks: off}
```
Comment (Araq) [inline lib/pure/bitops.nim]: Why is this necessary?

**Comment by Parashurama** [comment]
This was a workaround for an old issue I had with __builtin_ffs which only accepted cint arg, with overflew with values > high(cint). Also, I couln't lie about arg type in func definition with clang compiler.

**Comment by Araq** [comment]
Sorry it took so long.
